Allow modules and classes to specify required ancestors

Morriar opened the pull request:
### Motivation

(This coming from the documentation included in this pull-request)

It's not uncommon in Ruby to define helper modules that depends on other modules.
For example, let's take the following helper which provides `say_error` method:

```ruby
# typed: true

module MyHelper
  def say_error(message):
    raise "InternalError: #{message}" # error: Method `raise` does not exist on `MyHelper`
  end
end

class MyClass
  include MyHelper

  def do_something(x)
    say_error("some error") unless x
    # ...
  end
end
```

If we run Sorbet on this example, we will get a type-checking error saying that the method `raise`
does not exist on `MyHelper` since this method is defined on `Kernel`.
Thanks to this error, Sorbet is protecting us against some edge-cases where we would try to include
the `MyHelper` module in a class that does not include `Kernel`:

```ruby
class MyBaseClass < BasicObject
  include MyHelper

  def do_something(x)
    say_error("some error") unless x
    # ...
  end
end

MyBaseClass.new.do_something(false) # runtime-error: in `say_error': undefined method `raise' for #<MyBaseClass> (NoMethodError)
```

This example would raise an error at runtime because the method `raise` is undefined for instances of `MyBaseClass`
as it doesn't include `Kernel`.

## Requiring Ancestors

Sorbet provides the `requires_ancestor` method as a way to ensure that classes or modules including `MyHelper`
will also include `Kernel`.

Let's change our base example to use `requires_ancestor`:

```ruby
module MyHelper
  extend T::Helpers

  requires_ancestor Kernel

  def say_error(message):
    raise "InternalError: #{message}"
  end
end
```

This way we specify that any module including `MyHelper` must also include `Kernel`
and Sorbet will display an error if it's not the case:

```ruby
class MyBaseClass < BasicObject # error: `MyBaseClass` must include `Kernel` (required by `MyHelper`)
  include MyHelper
end
```

`requires_ancestor` also works to require that a specific class must be inherited:

```ruby
module MyHelper
  extend T::Helpers

  requires_ancestor Object

  def class_name
    self.class.name
  end
end

class MyBaseClass < BasicObject # error: `MyBaseClass` must inherit `Object` (required by `MyHelper`)
  include MyHelper
end
```

Note that requirements are transitive:

```ruby
class MyBaseClass2 < MyBaseClass # error: `MyBaseClass2` must inherit `Object` (required by `MyHelper`)
  include MyHelper
end
```

`requires_ancestor` can be used to require more than one ancestor:

```ruby
module Test
  module TestAssertions
    def assert_equal(x, y)
      x == y
    end
  end

  class TestBase
    def test; end
  end

  class TestCase < TestBase
    include TestAssertions
  end
end

module MyLogger
  def log_test_failed; end
end

module MyTestHelper
  extend T::Helpers

  requires_ancestor Test::TestAssertions, MyLogger

  def assert_not_equal(x, y)
    if assert_equal(x, y)
      true
    else
      log_test_failed
      false
    end
  end
end

class MyValidTest < Test::TestCase
  include MyTestHelper
  include MyLogger
end

class MyBrokenTest < Test::TestBase # error: `MyBrokenTest` must include `Test::TestAssertions` (required by `MyTestHelper`)
                                    # error: `MyBrokenTest` must include `MyLogger` (required by `MyTestHelper`)
  include MyTestHelper
end
```

### How it works?

1. A resolver job collects the required ancestors specified with `requires_ancestor X, Y` and stores them in a magic `<required-ancestor>` property.

2. The symbol can read the list of directly required ancestors from this magic `<required-ancestor>` property. Or compute recursively the list of all its ancestors on demand.

3. The definition validator will use both these lists to ensure all non-abstract classes do include all the ancestors their own includes require (and a few other sanity checks).

4. During calls validation, if a method called is not found, we also look into the required ancestors to see if the method exists before raising an error.

### But at what cost?

None if you don't use any `requires_ancestor` in your code.

If you do use them, you'll pay the price only where used:
* Cost some memory to store the magic "<required-ancestor>" property only for classes/modules that do use it.
* If a method is not found on a receiver, cost a bit to recursively look into all ancestors of the receiver if one of the required ancestors provides the method. So you pay only for NoMethod errors if you don't actually use the feature.

### Reviews

This pull-request can be read commit by commit.

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests for more examples on how it works and the kind of errors it produces.

This should also provide an answer to https://github.com/sorbet/sorbet/issues/2236.

jez commented on core/types/calls.cc:
```diff
@@ -536,7 +536,8 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, DispatchArgs args,
                 if (args.name == core::Names::declareInterface() || args.name == core::Names::declareAbstract() ||
-                    args.name == core::Names::declareFinal() || args.name == core::Names::declareSealed()) {
+                    args.name == core::Names::declareFinal() || args.name == core::Names::declareSealed() ||
+                    args.name == core::Names::requiresAncestor()) {
```
What's the solution when a module must `extend` a class, not `include` it?

jez replied:
Ah actually maybe this doesn't make sense because `self.` methods of a module are never mixed into the target module.

nvm!

Morriar replied:
We have a test to check this: https://github.com/sorbet/sorbet/pull/3409/files#diff-862cb4a86571973f989ed33b9bd48785R1-R21

jez replied:
yeah i was thinking the other way around, like

```ruby
module A
  def self.foo
    raise
  end
end
```

but this is already fine, with no need for the require_ancestor line.

jez commented on resolver/GlobalPass.cc:
```diff
@@ -345,20 +345,179 @@ void fullLinearizationSlowImpl(core::GlobalState &gs, const ParentLinearizationI
         }
     }
 };
+
```
I only just skimmed this code but couldn't find anything:

Does it make sense to use `requires_ancestor` inside a class? I think it only makes sense in in a module.

- If it makes sense in a class, is there a test?
- If it doesn't, can we emit an error for using it in a class, and write a test?

Morriar replied:
I added a test showing how to use it on a class: https://github.com/sorbet/sorbet/pull/3409/files#diff-deae14f8b7f473687b45d25ebbcbb820R1-R34

There are also two tests about errors with classes: https://github.com/sorbet/sorbet/pull/3409/files#diff-8619494c7637bf3655aac0a3eedfe6f0R33-R41

jez commented on rbi/sorbet/t.rbi:
```diff
@@ -87,7 +87,9 @@ module T::Helpers
   def sealed!; end
+
   def mixes_in_class_methods(mod); end
+  def requires_ancestor(*mod); end
```
Would you mind writing a signature for this? (Even if it's all untyped).

Morriar replied:
It would be the only method in that file with a `sig`. Should I create a separate PR to also add signatures to the other `T::Helpers` methods?

jez replied:
What do you mean? the `sealed!` and `final!` methods a few lines up also have sigs.

feel free to do this in a separate PR if you want.

Morriar replied:
Oh sorry I was confusing this file with `helpers.rb` from sorbet-runtime: https://github.com/sorbet/sorbet/pull/3409/files#diff-f82649c582c88ac321fcaa37ca936da2R27-R43 do we want a signature in this file?

Morriar replied:
Also, about adding a signature in `t.rbi`. It is the same problem than for `mixes_in_class_methods` and I believe it's why it doesn't have a signature either:

Since the checks for the `requires_ancestor` signature are done early in the `namer`, adding a signature would mean duplicating the errors during the `calls` checking.

For example:

```ruby
class A
  requires_ancestor "A"
end
```

Would return the error: 

```
error: `requires_ancestor` must only contain constant literals
```

As expected. But would also return:

```
error: Expected `Module` but found `NilClass` for argument `mod`
```

Because we discarded the node during the `namer` phase.

jez replied:
gotcha; i'm fine to just have a comment there saying why there's no sig, for both of them.

Morriar replied:
Good idea, I added a comment.

jez commented on resolver/resolver.cc:
```diff
@@ -121,7 +121,8 @@ class ResolveConstantsWalk {
         core::FileRef file;
 
-        bool isSuperclass; // true if superclass, false for mixin
+        enum Kind { superclass, mixin, requiredAncestor };
```
Can you use the same case that you'd use for class names here? Also, use `enum class` in C++. It's a lot better than `enum` for a handful of reasons.
```suggestion
        enum class Kind { Superclass, Mixin, RequiredAncestor };
```

djudd-stripe commented:
I've really wanted something like this! Not qualified to comment on the implementation, but a question about naming - this feels closely related to the `bind` method on `T.proc`, in that it's basically an assertion about the type of `self`. Should it have `bind` in the name somehow, like `bind_self`? I can't think of anything that's quite as clear as `required_ancestor`, though...

Morriar commented:
@djudd-stripe 

> in that it's basically an assertion about the type of `self`. Should it have `bind` in the name somehow, like `bind_self`?

Good point. My feelings against `bind` is that it doesn't give the user any clue on how to solve the problem. While `requires_ancestor` at least points to the including/subclassing mechanism.

I'm not opposed to change the name if you come with something better.

jez commented on gems/sorbet-runtime/lib/types/helpers.rb:
```diff
@@ -24,6 +24,23 @@ def sealed!
+  #     include MyHelper
+  #   end
+  #
+  # TODO: implement the checks in sorbet-runtime.
```
What's your plan for sorbet-runtime?

Morriar replied:
No, it does not require anything beside this definition. The requirement will just not be checked at runtime.

I think @paracycle was interested in providing the runtime behaviour in a follow-up PR.

jez commented on core/Symbols.h:
```diff
@@ -623,6 +654,9 @@ class Symbol final {
     InlinedVector<SymbolRef, 4> mixins_;
 
+    // Ancestors required by this class or module associated to the `requires_ancestor X` location
+    InlinedVector<Symbol::RequiredAncestor, 4> requiredAncestorsLocs;
```
This is a ton of memory to dedicate to something which is almost never used. This is going to make the Symbol table much, much larger. `RequiredAncestor` is 4 bytes, and you're reserving 4 of them inline.

In general, I think you're storing too much information. We don't store a `Loc` nor a `requiredBy` `SymbolRef` for `mixins_`. Why do we need to for `requiredAncestors`? Don't you have enough information to compute that information at the error site, like we do for `mixins_`? I think we should only need to store the `SymbolRef required;` field, which would mean we could delete the `RequiredAncestor` struct entirely.

jez replied:
On top of that, I would probably drop this down to `InlinedVector<..., 1>` because that's more likely to be the common case.

The way we found the numbers for every other `InlinedVector` in Sorbet before was to instrument Sorbet to dump a bunch of counters, and then burn those numbers in.

But for a new feature which is almost never used, it's not so much required. Maybe after a year of usage we could revisit it but a default of `1` is probably the best bet (most people are just going to do `Kernel` and call it a day).

Morriar replied:
* `requiredBy` is used to displayed which module required the ancestor (it might not be `self`). It's super useful to debug:

    ```
    B is missing X (required by Y)
    ```

    It's cached to make the error reporting faster but for this information I think I can go up the hierarchy and find the source.

* `requiredAt` is used to display the proper location of where `X` was required in `Y`. I believe that after the `resolver` this information is lost for ever. I think I used the same approach than for the `sealedLocs`: https://github.com/sorbet/sorbet/blob/master/core/Symbols.h#L522

We can limit the memory footprint but I fear it will be for less expressive errors.

jez replied:
I'm not convinced. We emit pretty much the same errors in `definition_validator` but we don't track the `Loc` information there.

In fact, did you ever look at implementing as much as possible of these error-emitting checks in `definition_validator`? Right now we're doing a huge loop over every `Symbol` in `GlobalState`. If these checks were in `definition_validator`, they would only run for the files that changed.

Morriar replied:
The whole thing is heavily inspired from the validator for `abstract` (you can see it with the caching mechanism actually). At first I put the checks in the `definition_validator` but I also need to change the GlobalPass to perform the linearization of the required ancestors (so method resolution works). So I moved everything to the class in the `GlobalPass` to minimize the diffs.

jez commented on resolver/GlobalPass.cc:
```diff
@@ -388,6 +545,14 @@ void Resolver::finalizeSymbols(core::GlobalState &gs) {
     computeLinearization(gs);
 
+    // Compute the lineazrize list of all required ancestors for each class or module
+    RequiredAncestorsLinearization requiredAncestorsLinearization(gs);
```
Why is this a class instead of a function? It looks like the only benefit is to avoid passing around `gs` to every helper function. We prefer threading that around explicitly (there is pretty much no class that takes a `GlobalState` and holds onto it).

Morriar replied:
I need it to hold a local state for the linearization cache: https://github.com/sorbet/sorbet/pull/3409/files/5fd178a1dfa0c3f1784a66a46adf9573ab32d6ff#diff-9645170eeb3775ef8bd3dac5fe8375b9R378-R379

I can add all of this in the `Resolver` if you prefer?

jez replied:
I see. In that case, I'd prefer keeping it a class, but still taking the `GlobalState &` argument by manually threading it through all the call sites.

Morriar commented:
I looked at the failing error for LSP. It seems that the errors are not reported again after the fast path.

I added a few prints to understand what is happening and the check is actually called and does the right thing. It's `beginError` that doesn't seem to display the error. Is it expected? 

Should I disable the fast path for those tests or is there a way to force `beginError` to  create the error a second time?

jez commented:
> I looked at the failing error for LSP. It seems that the errors are not reported again after the fast path.
> 
> I added a few prints to understand what is happening and the check is actually called and does the right thing. It's `beginError` that doesn't seem to display the error. Is it expected?
> 
> Should I disable the fast path for those tests or is there a way to force `beginError` to create the error a second time?

I think this indicates a somewhat fundamental limitation with the design. The fast path by design only runs the `runTreePasses` phase of resolver. This excludes most of the `GlobalPass` checks. This is part of what makes the fast path fast: it only re-runs Sorbet on the parts that it thinks could have changed, and all the GlobalPass logic is slow.

> The whole thing is heavily inspired from the validator for abstract (you can see it with the caching mechanism actually). At first I put the checks in the definition_validator but I also need to change the GlobalPass to perform the linearization of the required ancestors (so method resolution works). So I moved everything to the class in the GlobalPass to minimize the diffs.

I think what you want to do is move as much of your logic back to `definition_validator`, which is run per changed file, not for all files. When the fast path runs, the `definition_validator` checks will always run a second time, so the errors that are currently not showing up will start to show up.

I think that by moving the logic to `definition_validator`, you will also be forced to stop using the `ancestorsCache` that you're currently using. What was your goal with that? It's called `linearizeRequiredAncestors` but it doesn't seem to be doing something similar to what `computeLinearization` is doing.

jez commented:
> I've really wanted something like this! Not qualified to comment on the implementation, but a question about naming - this feels closely related to the `bind` method on `T.proc`, in that it's basically an assertion about the type of `self`. Should it have `bind` in the name somehow, like `bind_self`? I can't think of anything that's quite as clear as `required_ancestor`, though...

@djudd-stripe When I see `bind` I think "this is some lurky `instance_eval` / `instance_exec` hack" and that's not what's going on at all. When we were discussing this feature internally, we always described it by comparison with abstract classes. The idea was that if you're using this feature, the class or module is essentially `abstract!` until it's required ancestors are met.

I don't have strong thoughts on the name. I think the API will always be something akin to `some_method_name SomeModule` which is super easy to write an autocorrect for later if we want to change the name.

jez commented:
Ok another follow up:

I chatted with @elliottt about this for about an hour trying to make sense of how this change works / should work.

I think we're both on the same page as far as moving as many of the `check*Ancestors` helpers into definition_validator, which should fix the LSP fast path test failures you're seeing.

Two other things that jumped out to us as out of place:

- The change to `findMemberTransitiveInternal` seems dangerous. It surprised us to see that code changing. Another change which I think has the same effect: there's a line in `builder_entry.cc` that ascribes the type of the synthetic `<self>` variable:

  https://github.com/sorbet/sorbet/blob/master/cfg/builder/builder_entry.cc#L34-L37

  I think for methods whose owning class has a non-empty `requiredAncestors()` list, we can change this `Cast` instruction to ascribe an intersection type, like `T.all(MyOwner, T.all(... all my required ancestors ...))`. So for a concrete example it might look like `<self> = T.cast(<self>, T.all(MyModule, Kernel))`. The intersection type would then take care of looking for members in different places, and we don't have to worry about changing a really low level primitive.

- The implementation of `linearizeRequiredAncestors` looks like it's doing something really similar to `computeLinearization`. `computeLinearization` already deals with the `superClass` and the `mixins` vectors on a Symbol, and it seems like `requiredAncestors` is similar in spirit to those. In particular, the way that method is written effectively uses the symbol table itself as a cache. There's an `isClassOrModuleLinearizationComputed` flag on every `ClassOrModule` symbol that says whether linearization was computed or not. That's kind of like your `alreadySeen` vector, and the symbol table itself is the `ancestorsCache`.

  Do you think you could try moving the logic to linearize required ancestors into `computeLinearization`? For example, this function has kind of two parts:

  https://github.com/sorbet/sorbet/blob/17bd9ff63dfb471574060acaf3e068e756f88cf9/resolver/GlobalPass.cc#L275-L326

  There's a short part at the top that handles the `superClass()` case. Then there's a longer part that handles the `mixins()` case. And I think we could add a third one that handles the `requiredAncestors` case.

  A couple benefits of this:
  - computeLinearization is already called from the `resolver::runTreePasses` function. Given the current state of the change, I think that we will incorrectly take the fast path when we should take the slow path, because the `linearizeRequiredAncestors` function is not called on the fast path. The way you can test this is to write an `# assert-slow-path: true` test (grep for some examples), like this:

  ```ruby
  # -- test/testdata/lsp/foo.rb --
  # typed: true
  # assert-slow-path

  module A
    def foo_a; end
  end
  module B
    # requires_ancestor A
    def foo_b
      foo_a
    end
  end
  ```


  ```ruby
  # -- test/testdata/lsp/foo.1.rbupdate --
  # typed: true
  # assert-slow-path

  module A
    def foo_a; end
  end
  module B
    requires_ancestor A
    def foo_b
      foo_a
    end
  end
  ```

  and this test will assert that the slow path must have been taken. This is important, because the symbol table changed, which means we have to type check everything again.

  - by sharing the traversal of GlobalState with the `mixins` and `superClass` in `computeLinearization`, it should have better cache locality, and should use less memory because we're re-using the symbol table to store things, instead of making `ancestorsCache`, which will be a huge helper data structure.

jez commented:
Sorry for the massive comments here. If you want to video chat some time, please don't hesitate to ask. I'm free almost any time from 10:00 a.m. – 6:00 p.m. PT.

Morriar commented:
I pushed a few changes that make everything work with both slow/fast path and I will try to address all the comments here.

First, I changed the implementation in several ways

* In the symbol, we only store the location of the locally required ancestors (vs the location of all linearized ancestors before)
* The error reporting traverses the ancestors recursively to find the first one which introduced the ancestor locally and get the location (a bit slower here because we can't rely on the linearization and have to traverse the tree)
* I moved all the checks to the validator pass
* What was missing was clearing the local ancestor list between reload thus ending in the same ancestor being required and added again to the symbol at reload

About the feature, I imagine the `requires_ancestor X` as two properties:
1. it **does** behave like `include X` to resolve the methods (methods from `X` are available from `self`)
2. it **does not** behave like `include X` for subtyping (`self` is not a subtype of `X`)

What I understand from the two solutions you are proposing would mean considering `X` as a parent of `self`.

My approach with this PR is the following:

1. In the resolver, when the file is (re)loaded, we clear the list of local ancestors and load all of them from the tree (they might have beed edited since the last time)
2. In the global pass, once everything is loaded we (re)compute the linearization of the ancestors for everything (since an ancestor might have been introduced somewhere we have to invalidate everything and start over).
3. In the validator. For each ancestor `X` in the linearized list of `self` (so all local + transitive ones but already linearized, not recursive) we check if `self` subtypes `X` (so the requirement has been satisfied somewhere in the inheritance tree).
4. In the Symbol when we're looking for a method, if it cannot be found anywhere, we look into the linearized list of ancestors (it's not recursive, it calls `findMember`). What I think is missing here is a case for when the linearization of the ancestors has not been computed yet, but I'm not sure how to test that.

I think the linearization of the ancestors must be based on the linearization of the mixins (and thus done after it) for multiple reasons:
* It's faster since all the mixins have been resolved for `self` we don't need to recurse for them. We only need to do it for the superclasses (unless we have that linearization somewhere too??) and for the required ancestors (because the requirements are transitive and we are building the said linearization for them).
* It allows us to break cycles and reflexive requirements. I think this is a necessary feature for the sake of the user + we can easily do it there.
* It's less intrusive, and avoid changing complex parts.

@jez regarding your example I added it to this PR to show that it works.

jez commented on resolver/resolver.cc:
```diff
@@ -605,7 +615,19 @@ class ResolveConstantsWalk {
+
+        // Each time the classdef is reloaded we need to clear the list of requiredAncestors
+        // before loading them again.
+        auto sym = klass.data(ctx);
+        sym->requiredAncestorsLocs.clear();
```
This doesn't look right. Deleting things from the symbol table is almost never what you want.

I think this breaks with a setup like this:

```ruby
# -- a_b.rb --
class A; end
class B; end

# -- m.rb --
module M
  requires_ancestor A
end

# -- n.rb --
module N
  requires_ancestor B
end
```

and then providing an edit to `m.rb` to add a blank line at the end. That would trigger the fast path, but all the `requiredAncestorLocs` would have been cleared, and only `A` will be added back as a required ancestor.

jez commented:
https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20MyAbstractModel%0A%20%20def%20query(query_string)%3B%20end%0Aend%0A%0Amodule%20ModelMixin%0A%20%20requires_ancestor%20MyAbstractModel%0A%20%20%23%20include%20MyAbstractModel%0A%0A%20%20%23%20ancestors%20%3D%20%5B...%2C%20%3CStartOfRequiredAncestors%3E%2C%20...%5D%0A%0A%20%20sig%20%7Bparams(...).returns()%7D%0A%20%20def%20%3Crequired_ancestors%3E%0A%20%20end%0A%0A%20%20def%20mixin_method%0A%20%20%20%20this%20%3D%20T.cast(self%2C%20T.all(ModelMixin%2C%20MyModel))%0A%20%20%20%20this%20%3D%20self%0A%20%20%20%20this.query('foo')%20%20%20%20%20%20%20%20%20%20%20%20%23%20why%20allow%20this%20but...%0A%20%20%20%20ModelHelpers.query_foo(this)%20%23%20...%20disallow%20this%3F%0A%0A%20%20%20%20super%0A%20%20end%0Aend%0A%0Aclass%20ModelHelpers%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams(model_instance%3A%20MyAbstractModel).void%7D%0A%20%20def%20self.query_foo(model_instance)%0A%20%20%20%20model_instance.query('foo')%0A%20%20end%0Aend%0A%0Aclass%20MyModel%20%3C%20MyAbstractModel%0A%20%20include%20ModelMixin%0Aend%0A%0AMyModel.new.mixin_method

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20Module%0A%20%20include%20T%3A%3ASig%0Aend%0A%0A%23%20--%20foo.rb%20--%0Amodule%20Parent%0A%20%20extend%20T%3A%3AHelpers%0A%20%20%23%20(1)%20Mark%20module%20as%20sealed%0A%20%20sealed!%0Aend%0A%0A%23%20(2)%20CAN%20be%20included%2Fextended%2Finherited%20in%20%60foo.rb%60%0Aclass%20Child1%3B%20include%20Parent%3B%20end%0Aclass%20Child2%3B%20include%20Parent%3B%20end%0A%0Asig%20%7Bparams(x%3A%20Parent).void%7D%0Adef%20foo(x)%0A%20%20case%20x%0A%20%20when%20Child1%20then%20%23%20...%0A%20%20when%20Child2%20then%20%23%20...%0A%20%20%23%20(3)%20Sealed%20classes%20support%20exhaustiveness%0A%20%20else%20T.absurd(x)%0A%20%20end%0Aend%0A%0A%0A%23%20--%20bar.rb%20--%0A%23%20(4)%20CANNOT%20be%20included%2Fextended%2Finherited%20in%20other%20files%0A%23%20class%20BadChild%3B%20include%20Parent%3B%20end%20%23%20error!

https://sorbet.run/#%23%20typed%3A%20strict%0Amodule%20MyAbstractModel%0A%20%20def%20query(query_string)%3B%20end%0Aend%0A%0Amodule%20ModelMixin1%0A%20%20include%20MyAbstractModel%20%23%20isClassOrModuleModule%0A%0A%20%20def%20mixin_method1%0A%20%20end%0Aend%0A%0Amodule%20ModelMixin2%0A%20%20include%20MyAbstractModel%0A%0A%20%20def%20mixin_method2%0A%20%20end%0Aend%0A%0Aclass%20MyModel%0A%20%20include%20ModelMixin1%0A%20%20include%20ModelMixin2%0A%20%20include%20MyAbstractModel%0Aend

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%3B%20end%0Amodule%20M%0A%20%20include%20A%0Aend

bbugh commented:
We have a similar problem to #2236. Our app has objects that have hundreds of calculated values in methods. Since there's so many, we've broken them out into separate modules for organization. Each class without separate modules would be thousands of lines long. 

```ruby
class SomeObject
  include SomeObjectCalc1
  include SomeObjectCalc2

  def some_object_method
  end
end

module SomeObjectCalc1
  def do_math
    some_object_method * 50
  end
end

module SomeObjectCalc2
  def do_another_math
    some_object_method * 100
  end
end
```

In this case, our required "ancestor" is the class it will be included in. 

Will this PR support this scenario? There doesn't seem to be any other way to model this in Sorbet and I've been watching for something like this PR (thanks!)

Morriar commented:
@jez new version of this PR 🎉 

First, I totally changed the approach so it does not require changes to the method resolution mechanism. Instead, we wait for `callDispatch` to *not* find a method called and check if a required ancestor provides it and if it does, we dispatch the call to this one. So we pay the price only in cases where Sorbet would raise a `NoMethod` error. The cost being looking if the superclasses and the linearized list of included modules (already computed done by Sorbet earlier) have required ancestors.
* If no required ancestor is found for the receiver type, `dispatchCall` raises the expected `NoMethod` error.
* If there are transitively required ancestors for the receiver type, we try to dispatch the call to each of them until the first one that accepts it, or raise the `NoMethod` error if none can.

Then, I also changed the way the data are stored to use a fake method definition like it's done for sealed classes. Ancestors are stored in the return type (a `TupleType`) and the associated locations in the method symbol `locs` so the memory footprint is 0 if you don't use the feature.

Finally, to answer the remaining questions in this thread:

* Can we reuse the module linearization process?
  No since we allow multiple required ancestors and ancestors to be (abstract) classes, so it wouldn't fit inside the current module linearization process. Also, no since we allow requirement cycles. And finally, no since creating a requirement on an ancestor does not mean creating a subtyping relationship between the requiree and the required. It only creates a constraint on the thing including / subclassing the requiree. 

  Note that we still reuse the module linearization as it was computed in the GlobalPass when going up the tree so this part is not duplicated.

* What about `sorbet-runtime`?
  Currently, `requires_ancestor` just no-ops at runtime. So the runtime works as before and will raise the same error as before is something required should have been included and is not.

* Is it supported in LSP mode?
  Yes, see the test under `test/testdata/lsp` and also not the the POS tests do not diable the fast path.

Morriar commented on core/Symbols.cc:
```diff
@@ -1112,6 +1112,116 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+    auto ancestors = this->findMember(ctx, Names::requiredAncestors());
+    if (!ancestors.exists()) {
+        ancestors = ctx.state.enterMethodSymbol(loc, this->ref(ctx), Names::requiredAncestors());
+        ancestors.data(ctx)->locs_.clear(); // Remove the original location
```
`ctx.state.enterMethodSymbol(loc, this->ref(ctx), Names::requiredAncestors())` stores the `loc` in the Symbol's `locs` array. Let's remove it to avoid some offsetting later.

Morriar commented on core/Symbols.cc:
```diff
@@ -1112,6 +1112,116 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+
+    if (ancestors.data(ctx)->resultType == nullptr) {
+        vector<TypePtr> targs;
+        ancestors.data(ctx)->resultType = TupleType::build(ctx, move(targs));
```
We store the ancestors as the typle `[ANCST1, ANCST2...]`

Morriar commented on core/Symbols.cc:
```diff
@@ -1112,6 +1112,116 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+
+        RequiredAncestor req;
+        req.symbol = cast_type_nonnull<ClassType>(elem).symbol;
+        req.loc = data->locs().at(index);
```
The loc the ancestor was required at (ie the loc of the `requires_ancestor X`) is at the same index than the `ANCST` type in the returned tuple.

Morriar commented on core/Symbols.cc:
```diff
@@ -1112,6 +1112,116 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+    std::vector<Symbol::RequiredAncestor> res;
+
+    if (std::find(seen.begin(), seen.end(), sym) != seen.end()) {
+        return res; // Break recursive loops if we already visited this ancestor
```
So we avoid infinite loops in circular requirements.

Morriar commented on core/Symbols.cc:
```diff
@@ -1112,6 +1112,116 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+        }
+    }
+
+    return res;
```
Ideally we should cache this result the first time it's called, but we not might have a mutable context at the time we will call this method (ie in `calls.cc`).

Morriar commented on core/Symbols.h:
```diff
@@ -582,6 +582,22 @@ class Symbol final {
+
+    // Associate a required ancestor with the loc it's required at
+    struct RequiredAncestor {
+        SymbolRef origin; // Only used during transitive lookups
```
We reuse the same struct for `requiredAncestors` and `requiredAncestorsTransitive`. In the former, the `origin` is always `this`. In the later, it can be any ancestor of `this`.

Morriar commented on core/tools/generate_names.cc:
```diff
@@ -275,6 +275,10 @@ NameDef names[] = {
     {"attached", "<attached class>"},
 
+    // Requires ancestor
+    {"requiredAncestors", "<required-ancestor>"},
```
For the fake, magic method that will hold the list.

Morriar commented on core/types/calls.cc:
```diff
@@ -518,6 +518,17 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
 
+    if (!mayBeOverloaded.exists()) {
+        // Before raising any error, we look if the method exists in all required ancestors by this symbol
+        auto ancestors = symbol.data(gs)->requiredAncestorsTransitive(gs);
```
Only is the method is not found, this part is where we pay the most perf:

We look into all transitive superclasses (recursive like when we look with `findMemberTransitive`) and in the linearized list of mixins if they require an ancestor.

Morriar commented on core/types/calls.cc:
```diff
@@ -560,7 +571,7 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
                 if (args.name == core::Names::declareInterface() || args.name == core::Names::declareAbstract() ||
                     args.name == core::Names::declareFinal() || args.name == core::Names::declareSealed() ||
-                    args.name == core::Names::mixesInClassMethods()) {
+                    args.name == core::Names::mixesInClassMethods() || args.name == core::Names::requiresAncestor()) {
```
Just so we can suggest include `T::Helpers`

elliottt replied:
```suggestion
                    args.name == core::Names::mixesInClassMethods() || (args.name == core::Names::requiresAncestor() && gs.requiresAncestorEnabled)) {
```

Morriar commented on definition_validator/validator.cc:
```diff
@@ -597,10 +680,12 @@ class ValidateWalk {
         validateAbstract(ctx, singleton);
         validateFinal(ctx, sym, classDef);
         validateSealed(ctx, sym, classDef);
+        validateRequiredAncestors(ctx, singleton);
```
We also check the singleton for when something `extend` a module with required ancestors.

Morriar commented on resolver/resolver.cc:
```diff
@@ -595,6 +608,46 @@ class ResolveConstantsWalk {
         owner.data(gs)->members()[core::Names::classMethods()] = id->symbol;
     }
 
+    static void resolveRequiredAncestorsJob(core::GlobalState &gs, const RequireAncestorResolutionItem &todo) {
```
This part is very very similar to `mixin_in_class_module`

elliottt commented on resolver/resolver.cc:
```diff
@@ -169,11 +169,24 @@ class ResolveConstantsWalk {
+        core::SymbolRef owner;
+        ast::Send *send;
+
+        RequireAncestorResolutionItem(RequireAncestorResolutionItem &&) noexcept = default;
+        RequireAncestorResolutionItem &operator=(RequireAncestorResolutionItem &&rhs) noexcept = default;
+
+        RequireAncestorResolutionItem(const RequireAncestorResolutionItem &) = delete;
+        const RequireAncestorResolutionItem &operator=(const RequireAncestorResolutionItem &) = delete;
```
Were you running into copying problems that this resolved?

Morriar replied:
No problem at all but all other `*ResolutionItem` have it. I can remove it if you prefer.

elliottt commented on resolver/resolver.cc:
```diff
@@ -609,6 +609,39 @@ class ResolveConstantsWalk {
+            return;
+        }
+
+        if (send->args.size() < 1) {
```
I'd use `send->args.empty()` instead here.

elliottt commented on core/Symbols.cc:
```diff
@@ -1112,6 +1112,65 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+        ancestors.data(ctx)->resultType = TupleType::build(ctx, move(targs));
+    } else {
+        // Do not require the same ancestor twice
+        for (auto elem : (cast_type<TupleType>(ancestors.data(ctx)->resultType))->elems) {
+            ENFORCE(isa_type<ClassType>(elem), "Something in requiredAncestors that's not a ClassType");
+            if (cast_type_nonnull<ClassType>(elem).symbol == ancestor) {
+                return;
+            }
+        }
```
It might be worth trying to rephrase this in terms of `absl::any_of, just to clarify the intent.

Morriar replied:
I changed it, and I'm not entirely sure it's clearer:

![image](https://user-images.githubusercontent.com/583144/102410778-0891ac00-3fbf-11eb-9122-cb14a33564ea.png)

For me, the comment just above is the more helpful, I let you judge.

elliottt commented on core/Symbols.cc:
```diff
@@ -1112,6 +1112,65 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+        RequiredAncestor req;
+        req.symbol = cast_type_nonnull<ClassType>(elem).symbol;
+        req.loc = data->locs().at(index);
+        res.emplace_back(req);
```
You can construct in-place with `emplace_back`, so `res.emplace_back(cast_type..., data->locs...);` will work if there's a two argument constructor for `RequiredAncestor`.

elliottt commented on core/Symbols.h:
```diff
@@ -582,6 +582,18 @@ class Symbol final {
+    struct RequiredAncestor {
+        SymbolRef symbol;
+        Loc loc;
+    };
```
I'd add a constructor for `RequiredAncestor` to make working with vectors easier.

elliottt commented on core/Symbols.cc:
```diff
@@ -1112,6 +1112,65 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
 }
 
+// Record a required ancestor for this class of module
+void Symbol::recordRequiredAncestor(MutableContext ctx, SymbolRef ancestor, Loc loc) {
```
Does this need to be `MutableContext`? It doesn't look like there's anything that requires more than `GlobalState &` in this method.

Morriar replied:
Indeed, a non-const `GlobalState` is enough

elliottt commented on definition_validator/validator.cc:
```diff
@@ -500,6 +500,23 @@ void validateSealed(core::Context ctx, const core::SymbolRef klass, const ast::C
+
+void validateRequiredAncestors(core::Context ctx, const core::SymbolRef sym) {
+    validateUselessRequiredAncestors(ctx, sym);
+}
```
It would be nice if this function could exit early if there were no required ancestors present on the symbol.

Morriar replied:
We would also need to check the transitive ones to ensure all requirements are satisfied

elliottt replied:
Could you write up an example for me?

Morriar replied:
With all commits considered, `validateRequiredAncestors` does the following:

```cpp
void validateRequiredAncestors(core::Context ctx, const core::SymbolRef sym) {
    validateUselessRequiredAncestors(ctx, sym);
    validateUnsatisfiedRequiredAncestors(ctx, sym);
    validateUnsatisfiableClassAncestors(ctx, sym);
}
```

Where:

* `validateUselessRequiredAncestors` indeed needs only the local required ancestors
* `validateUnsatisfiedRequiredAncestors` needs the transitive ones (but early returns for modules and abstract classes)
* `validateUnsatisfiableClassAncestors` needs the required classes from the transitively required ancestors

Morriar replied:
So for something like

```rb
module RA; end

module M
  extend T::Helpers
  requires_ancestor RA
end

class C1
  extend T::Helpers
  include M
end

class C2 < C1 # error: Should include RA
end
```

Even if `C2` has no locally required ancestors, we need to look in it's superclasses and mixins (`requiredAncestorsTransitive`) if one of them requires an ancestors to be able to display the error.

elliottt commented on core/Symbols.cc:
```diff
@@ -1171,6 +1171,57 @@ vector<Symbol::RequiredAncestor> Symbol::requiredAncestors(const GlobalState &gs
+                                                                     std::vector<SymbolRef> seen) {
+    std::vector<Symbol::RequiredAncestor> res;
+
+    if (std::find(seen.begin(), seen.end(), sym) != seen.end()) {
```
Could you use `absl::c_find(seen, sym)` instead? It makes it easier to avoid common problems with iterators.

elliottt commented on core/Symbols.cc:
```diff
@@ -1171,6 +1171,57 @@ vector<Symbol::RequiredAncestor> Symbol::requiredAncestors(const GlobalState &gs
 
+// All required ancestors by this class or module
+vector<Symbol::RequiredAncestor> requiredAncestorsTransitiveInternal(const GlobalState &gs, SymbolRef sym,
+                                                                     std::vector<SymbolRef> seen) {
```
Should `seen` be a reference? If not, this is copying the entire vector for each recursive call.

elliottt commented on core/Symbols.cc:
```diff
@@ -1171,6 +1171,57 @@ vector<Symbol::RequiredAncestor> Symbol::requiredAncestors(const GlobalState &gs
+// All required ancestors by this class or module
+vector<Symbol::RequiredAncestor> Symbol::requiredAncestorsTransitive(const GlobalState &gs) const {
+    ENFORCE(this->isClassOrModule(), "Symbol is not a class or module: {}", this->show(gs));
+    std::vector<SymbolRef> seen;
+    return requiredAncestorsTransitiveInternal(gs, this->ref(gs), seen);
```
With a little bit of refactoring, isn't `seen` basically what you're trying to return? You could thread it through as a reference, and use `absl::find_if` with a lambda to project the values out of `seen`.

Morriar replied:
Not sure how you would do it.

Be the following example:

```rb
module RA; end

module M
  requires_ancestor RA
end

class C
  include M
end
```

`seen` would contain `C, M, RA` (ie all the thing we *traversed*) while `res` would contain only `RA` (ie all the thing we *collected*).

Am I missing something?

elliottt replied:
I see. Well, my comment still stands about copying with `seen`. You could thread references to both the seen vector, and the `res` vector through, and then you wouldn't end up re-allocating/copying quite so much.

Morriar replied:
I've already done it for `seen`, but yeah, good idea about `res`, I'll change this.

elliottt commented on definition_validator/validator.cc:
```diff
@@ -530,9 +530,57 @@ void validateUnsatisfiedRequiredAncestors(core::Context ctx, const core::SymbolR
+    auto data = sym.data(ctx);
+
+    vector<core::Symbol::RequiredAncestor> requiredClasses;
+    for (auto ancst : data->requiredAncestorsTransitive(ctx)) {
```
This is somewhat expensive to compute, why not compute it once up front in `validateRequiredAncestors` and pass it through as a const reference?

Morriar replied:
I would be happy to precompute it once the share it, but we also need it from `calls`. Any idea?

Morriar replied:
Done in the new version at the end of the global pass 👍

elliottt commented on definition_validator/validator.cc:
```diff
@@ -530,9 +530,57 @@ void validateUnsatisfiedRequiredAncestors(core::Context ctx, const core::SymbolR
+        auto ra1 = requiredClasses[i];
+        for (int j = i + 1; j < requiredClasses.size(); j++) {
+            auto ra2 = requiredClasses[j];
+            if (ra1.symbol != ra2.symbol && !ra1.symbol.data(ctx)->derivesFrom(ctx, ra2.symbol) &&
```
Isn't it alway the case that `ra1.symbol != ra2.symbol`, as `requiredAncestorsTransitive` returns a vector of unique symbols?

Morriar commented:
@elliottt I handled you comments and did a few improvements:

* The transitive list of ancestors is computed ahead of time during the global pass and saved inside a magic `<required-ancestors-lin>` property which stores the list of required ancestors for a symbol, the origin or the requirement, and the loc of the requirement.
* No more `res` passed around, the ancestors are directly recorded inside the `<required-ancestors-lin>` property.

Morriar commented on resolver/GlobalPass.cc:
```diff
@@ -402,6 +402,9 @@ void Resolver::finalizeSymbols(core::GlobalState &gs) {
         resolveTypeMembers(gs, sym, typeAliases, resolved);
+
+        // Precompute the list of all required ancestors for this symbol
+        sym.data(gs)->computeRequiredAncestorLinearization(gs);
```
Should I clear the cache before?

I was expecting this to fail in LSP tests but adding / removing ancestors in `test/testdata/lsp/requires_ancestor.*.rbupdate` seems to work properly, without cash clearing.

Morriar replied:
No, this is where we fill the cache

Morriar commented:
Rebased on master

elliottt commented:
Here's the example we talked about today @Morriar:

```ruby

# typed: true
module A
  def bar(x); end
end
module B
  def bar; end
end
module C
  requires_ancestor A
  requires_ancestor B
  def foo
    bar
  end
end
class Parent
  include B
  include A
end
class Child < Parent
  include C
  def main
    foo
  end
end
Child.new.main
```

Morriar commented:
I protected the feature behind a `--enable-experimental-requires-ancestor` flag

elliottt commented on core/types/calls.cc:
```diff
@@ -517,6 +517,17 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
 
     SymbolRef mayBeOverloaded = symbol.data(gs)->findMemberTransitive(gs, args.name);
 
+    if (!mayBeOverloaded.exists()) {
```
```suggestion
    if (!mayBeOverloaded.exists() && gs.requiresAncestorEnabled) {
```

elliottt commented on core/Symbols.cc:
```diff
@@ -1165,6 +1165,143 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+// * RequiredAncestor.origin goes into the first argument type tuple
+// * RequiredAncestor.loc goes into the symbol loc
+// All fields for the same RequiredAncestor are stored at the same index.
+void Symbol::recordRequiredAncestorInternal(GlobalState &gs, Symbol::RequiredAncestor &ancestor, NameRef prop) {
```
Could this method and `readRequiredAncestorsInternal` be functions defined in an anonymous namespace? Instead of relying on `this`, they could take a `Symbol *` argument called `self`.

Morriar replied:
The problem is accessing `_locs` so I can `clear` it from the location we add by default.

elliottt replied:
I was wondering after I left this comment if it was missing because I asked you to remove it, thanks! 😅

elliottt commented on core/Symbols.cc:
```diff
@@ -1165,6 +1165,143 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+
+// All required ancestors by this class or module
+vector<Symbol::RequiredAncestor> Symbol::requiredAncestorsTransitive(const GlobalState &gs) const {
+    return readRequiredAncestorsInternal(gs, Names::requiredAncestorsLin());
```
```suggestion
    ENFORCE(gs.requiresAncestorEnabled);
    return readRequiredAncestorsInternal(gs, Names::requiredAncestorsLin());
```

elliottt commented on core/Symbols.cc:
```diff
@@ -1165,6 +1165,143 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+    return readRequiredAncestorsInternal(gs, Names::requiredAncestorsLin());
+}
+
+void Symbol::computeRequiredAncestorLinearization(GlobalState &gs) {
```
```suggestion
void Symbol::computeRequiredAncestorLinearization(GlobalState &gs) {
    ENFORCE(gs.requiresAncestorEnabled);
```

elliottt commented on test/testdata/lsp/requires_ancestor.rb:
```diff
@@ -0,0 +1,18 @@
+# typed: true
+# enable-experimental-requires-ancestor: true
```
Is this used by the test suite?

Morriar replied:
Yes, for LSP test runner: https://github.com/sorbet/sorbet/pull/3409/files#diff-9a281b58f3ec5ea0244c2ac093f3298a22f1fe00828e2ff2e4cefb468ac94ebeR323

I also added them to the other tests so the `pipeline_test_runner` enables the option only when needed :+1

elliottt reviewed (changes requested):
Thanks for putting this feature behind a flag!

I added checks for the feature flag in a few place, and a few ENFORCEs to make sure that the behavior is only triggered when the flag is enabled. There were a few cases where it would take more than a suggestion comment to add the flag guard, could you take a look at those?

bbugh commented:
I asked this before, but I wanted to give a specific use case to see if this will be supported by this branch. We have hundreds of extra methods for our models and have organized them into related modules. We need to be able to call into other modules and the class' instance methods, as it's only for organizational purposes.

```ruby
class SomeModelClass
  include SomeModelClassExtension 
  include SomeModelClassExtension2
  include SomeModelClassExtension3
  include SomeModelClassExtension4
  include ...

  def project
  end
end

module SomeModelClassExtension
  requires_ancestor SomeModelClass # ???
  include_types SomeModelClass # ???

  def some_activity
    project # Sorbet currently does not seem to be able to type this
  end
end
```

Thanks!

Morriar commented on resolver/resolver.cc:
```diff
@@ -786,9 +838,21 @@ class ResolveConstantsWalk {
+            if (send.fun == core::Names::mixesInClassMethods()) {
+                auto item = ClassMethodsResolutionItem{ctx.file, ctx.owner, &send};
+                this->todoClassMethods_.emplace_back(move(item));
+            } else if (send.fun == core::Names::requiresAncestor()) {
+                if (!ctx.state.requiresAncestorEnabled) {
+                    if (auto e = ctx.beginError(send.loc, core::errors::Resolver::ExperimentalRequiredAncestor)) {
+                        e.setHeader("`{}` is an experimental feature. Use `{}` to enable it", "requires_ancestor",
+                                    "--enable-experimental-requires-ancestor");
+                    }
+                } else {
+                    auto item = RequireAncestorResolutionItem{ctx.file, ctx.owner, &send};
+                    this->todoRequiredAncestors_.emplace_back(move(item));
+                }
+            }
```
Note that the method will still be defined for Sorbet from it's RBI even without the flag, so you will be calling it but it will `no-op`. It might be better to explicitly error

That said, I removed the error as you asked.

Morriar commented on test/pipeline_test_runner.cc:
```diff
@@ -190,6 +190,7 @@ TEST_CASE("PerPhaseTest") { // NOLINT
         gs->semanticExtensions.emplace_back(provider->defaultInstance());
     }
 
+    gs->requiresAncestorEnabled = true;
```
I realized I can also read it from a `# enable-experimental-requires-ancestor: true` comment on top.

Morriar commented on main/lsp/wrapper.cc:
```diff
@@ -23,6 +23,8 @@ void setRequiredLSPOptions(core::GlobalState &gs, options::Options &options) {
         gs.suppressErrorClass(sorbet::core::errors::Namer::MultipleBehaviorDefs.code);
     }
 
+    gs.requiresAncestorEnabled = true;
```
Yeah we can read this from `options` instead 👍

Morriar commented:
@elliottt I implemented all the changes you required but the one about moving the method from private to anonymous as it's breaking current approach and ends up making the code harder to read.

It's ready for another review 👍

Morriar commented:
@bbugh: yes, the `requires_ancestor` will ensure that only `SomeModelClass ` (or a subclass) can include `SomeModelClassExtension ` making `project` available.

Not sure what is the `include_types` about?

bbugh commented:
@Morriar Awesome, thank you! And thank you for this fantastic branch! We'll be able to type our code that would benefit most from strong types. 

P.S. Sorry for the confusion about `include_types`, I couldn't remember the name of `requires_ancestor` and it looks like I forgot to delete it when I added the correct one.

elliottt commented on test/pipeline_test_runner.cc:
```diff
@@ -194,6 +194,10 @@ TEST_CASE("PerPhaseTest") { // NOLINT
 
     auto assertions = RangeAssertion::parseAssertions(test.sourceFileContents);
+
+    gs->requiresAncestorEnabled =
+        BooleanPropertyAssertion::getValue("enable-experimental-requires-ancestor", assertions).value_or(false);
```
👍

elliottt reviewed (approved):
This looks great, thank you for adding the additional guards I requested! Just two requests:

* Just to make sure that your flag in the test suite is working as expected, could you add a test that uses `requires_ancestor` but doesn't enable the feature?
* If you wouldn't mind rebasing this on master, I'll test it on the stripe codebase to make sure that we don't see a performance regression.

Morriar commented:
@elliottt thanks for the review!

> * Just to make sure that your flag in the test suite is working as expected, could you add a test that uses `requires_ancestor` but doesn't enable the feature?

Done, see `test/testdata/resolver/requires_ancestor_disabled.rb`

> * If you wouldn't mind rebasing this on master, I'll test it on the stripe codebase to make sure that we don't see a performance regression.

And done and fixed the conflicts.

elliottt commented on test/testdata/resolver/requires_ancestor_disabled.rb:
```diff
@@ -0,0 +1,44 @@
+module Test2
+  module M1
+    extend T::Helpers
+    requires_ancestor M2
```
Why doesn't this raise an error when the feature is disabled?

Morriar replied:
The problem is the definition is coming from the Helpers RBi.

I previously added a check to raise an error if the method was used without the option.

You asked me to remove it here, and I raised the same concerns two weeks ago https://github.com/sorbet/sorbet/pull/3409#discussion_r568922403.

elliottt replied:
I was wondering after I left this comment if it was missing because I had asked you to remove it, thanks! 😅

elliottt reviewed (changes requested):
I just noticed that `requires_ancestor` usage doesn't raise an error if the feature is disabled. Could you add that to the resolver?

elliottt commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_IvxFK1YO3JlRaF
→ https://go/builds/bui_IvxFVKOMdjbmM2

elliottt commented:
> We have a policy of testing changes to Sorbet against Stripe's codebase before
> merging them. I've kicked off a test run for the current PR. When the build
> finishes, I'll share with you whether or how it failed. Thanks!
> 
> Stripe employees can see the build results here:
> 
> → https://go/builds/bui_IvxFK1YO3JlRaF
> → https://go/builds/bui_IvxFVKOMdjbmM2

This was a successful build with no change in perf 🎉
